### DISCUSSION TITLE
Questions about using mixcr to build reference genes

### DISCUSSION THREAD

**Discussion by CoooooolM** [milaboratory/mixcr#1868] [Q&A] [answered]
Hello, I am using mixcr to construct reference genes of Chondrichthyes. The gene source is IMGT. In this process, I met "CommandLine $ExecutionException: Error while running the command buildLibrary Java lang. IllegalArgumentException: Gene conflict: U50607 ". How can I solve it? I have uploaded my command code and error to the attachment, thank you.
![7cdf3a8c-c94c-4c9b-ada0-1923fc347e64](https://github.com/user-attachments/assets/f0a8d1f0-b871-446e-9bef-9abb230924bf)
![f0696dbb-cdfe-4847-b1a9-d21bb74db5a6](https://github.com/user-attachments/assets/1390969f-e86e-47a1-8c17-d747ec28c00f)

**Comment by mizraelson** [comment, marked as answer]
To create a reference library, the sequences must be clearly defined. In your V genes FASTA file, all V genes contain ambiguous nucleotides in their sequences, which is not allowed for the reference.

For example:
```
>AF336094|IGHV2S8*01|Orectolobus maculatus|(F)|V-REGION| | | | | | | | | | |
acaagggtagaccaaacaccaagaacaNNNgcaacaaaagagacgggcgaatcactgacc
atcaactgcgtcctaagagatgctagttacNNNNNNNNNNNNggatcggagaggacgggc
tggtatcggacaaaattgggttcaacaNNNaacgagcagNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNacaatatcaattggcNNNggacgatatgttgaaacagtc
aacaagggatcaaagtccttttctctgagaattcgtgatctgagagttgaagacagtggc
acgtataagtgtggagca
```
​This is why you see the following message for all V genes:

Sequence dropped because contain wildcards: AF003853|IGHV1S1*01|Hydrolagus colliei|(F)|V-REGION| | | | | | | | | | |


​Additionally, the headers of the sequences need to be updated. MiXCR uses the first element of the header as the gene name. For example, in the following case:
```
>U50607|IGHJ4*01|Carcharhinus plumbeus|(F)|J-REGION
>U50607|IGHV1S5*01|Carcharhinus plumbeus|(F)|V-REGION
```
MiXCR will try to assign the same name (U50607) to both IGHJ4*01 and IGHV1S5*01, causing a conflict and resulting in an error.

You should change the headers to something like this:
```
>IGHJ4*01|Carcharhinus plumbeus|(F)|J-REGION
>IGHV1S5*01|Carcharhinus plumbeus|(F)|V-REGION
```
